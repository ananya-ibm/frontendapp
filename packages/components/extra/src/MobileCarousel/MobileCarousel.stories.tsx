/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import React from 'react';
import styled from 'styled-components';
import { ContentCard } from '@exo/frontend-components-content';
import { MobileCarousel } from './MobileCarousel';

export default {
  title: 'Components/Extra/MobileCarousel',
  component: MobileCarousel
};

const storyItems = [
  {
    title: 'New Offer 1',
    subtitle: 'Check out the latest offer',
    text: 'Isn\'t it a great offer. I bet you\'ve never seen one like it.',
    img:
      'https://images.unsplash.com/photo-1476286768413-e7051cdb2179?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=60',
    linkText: 'Learn More',
    link: '#',
    hasAlwaysShadow: true
  },
  {
    title: 'New Offer 2',
    subtitle: 'Check out the latest offer',
    text: 'Isn\'t it a great offer. I bet you\'ve never seen one like it.',
    img:
      'https://images.unsplash.com/photo-1476286768413-e7051cdb2179?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=60',
    linkText: 'Learn More',
    link: '#',
    hasAlwaysShadow: true
  },
  {
    title: 'New Offer 3',
    subtitle: 'Check out the latest offer',
    text: 'Isn\'t it a great offer. I bet you\'ve never seen one like it.',
    img:
      'https://images.unsplash.com/photo-1476286768413-e7051cdb2179?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=60',
    linkText: 'Learn More',
    link: '#',
    hasAlwaysShadow: true
  }
];

export const normal = args => (
  <MobileCarousel {...args}>
    {storyItems.map(item => (
      <div key={item.title} style={{ padding: '.5rem 0' }}>
        <ContentCard {...item} />
      </div>
    ))}
  </MobileCarousel>
);

export const withCustomSlideStyles = args => (
  <MobileCarousel {...args}>
    {storyItems.map(item => (
      <div key={item.title} style={{ padding: '.5rem 0' }}>
        <ContentCard {...item} />
      </div>
    ))}
  </MobileCarousel>
);

withCustomSlideStyles.args = {
  carouselStyles: {
    slideContainer: {
      padding: '0 3rem'
    }
  }
};

const Container = styled('div')`
  & .storybook-carousel {
    background: lightgrey;
    border: 0.0625rem solid black;
    padding: 1rem;
  }

  & .storybook-carousel-progress {
    justify-content: flex-start;

    & div {
      box-shadow: 0 0 0.5rem black;
    }
  }
`;

export const withCustomStyles = args => (
  <Container>
    <MobileCarousel {...args}>
      {storyItems.map(item => (
        <div key={item.title} style={{ background: 'white' }}>
          <ContentCard {...item} />
        </div>
      ))}
    </MobileCarousel>
  </Container>
);

withCustomStyles.args = {
  className: 'storybook-carousel'
};

export const withCustomProgressStyles = args => (
  <Container>
    <MobileCarousel {...args}>
      {storyItems.map(item => (
        <div key={item.title} style={{ padding: '.5rem 0' }}>
          <ContentCard {...item} />
        </div>
      ))}
    </MobileCarousel>
  </Container>
);

withCustomProgressStyles.args = {
  progressClassName: 'storybook-carousel-progress'
};

export const imageCarousel = args => (
  <MobileCarousel {...args}>
    {storyItems.map(item => (
      <img key={item.title} src={item.img} alt={item.title} />
    ))}
  </MobileCarousel>
);

imageCarousel.args = {
  carouselStyles: {
    slideContainer: {
      padding: '0'
    }
  }
};

export const withAutoPlay = args => (
  <MobileCarousel {...args}>
    {storyItems.map(item => (
      <div key={item.title} style={{ padding: '.5rem 0' }}>
        <ContentCard {...item} />
      </div>
    ))}
  </MobileCarousel>
);

withAutoPlay.args = {
  hasAutoPlay: true
};

export const withoutProgressButtons = args => (
  <MobileCarousel {...args}>
    {storyItems.map(item => (
      <div key={item.title} style={{ padding: '.5rem 0' }}>
        <ContentCard {...item} />
      </div>
    ))}
  </MobileCarousel>
);

withoutProgressButtons.args = {
  hasProgressButtons: false
};
