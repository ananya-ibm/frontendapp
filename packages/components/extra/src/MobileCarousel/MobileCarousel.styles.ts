/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import styled, { css } from 'styled-components';
import { ifProp, media } from '@exo/frontend-common-style-utils';
import theme from './MobileCarousel.theme';

export const MobileCarousel = styled('div')`
  ${props => media.greaterThan(props, 'medium').then(css`
    display: none;
  `)}
`;

export const Progress = styled('div')`
  display: flex;
  justify-content: center;
  padding: ${props => theme(props).progressPadding};
`;

export const Dot = styled('div')`
  background: ${props => theme(props).dotColor};
  border-radius: 50%;
  cursor: pointer;
  height: 60%;
  width: 60%;
`;

export const Button = styled('button')<{ isCurrent?: boolean }>`
  background: transparent;
  border: 0 none;
  cursor: pointer;
  height: ${props => theme(props).buttonSize};
  opacity: 0.4;
  transition: 200ms opacity ease-in;
  width: ${props => theme(props).buttonSize};
  z-index: 3;

  ${props => ifProp(props, 'isCurrent').then(css`
    opacity: 1;

    ${Dot} {
      background-color: ${theme(props).dotActiveColor};
    }
  `)}

  &:hover, &:focus {
    outline: none;
  }
`;
