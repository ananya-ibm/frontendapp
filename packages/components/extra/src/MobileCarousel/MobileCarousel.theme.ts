/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

type Props = {
  buttonSize: string;
  dotActiveColor: string;
  dotColor: string;
  progressPadding: string;
};

declare global {
  interface EXOComponentStyles {
    extra_mobileCarousel?: Props;
  }
}

export default (props: EXOThemeProps): Props => {
  return {
    buttonSize: '1.25rem',
    dotActiveColor: props.theme.colors.brand.brand1.base,
    dotColor: props.theme.colors.delimiters.highContrast,
    progressPadding: props.theme.spacing.inline.s3,

    ...props.theme?.byComponent?.extra_mobileCarousel
  };
};
