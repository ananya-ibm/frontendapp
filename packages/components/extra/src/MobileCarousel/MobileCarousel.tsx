/* eslint-disable react/jsx-indent-props */
/* eslint-disable react/jsx-indent */
/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import React from 'react';
import SwipeableViews from 'react-swipeable-views';
import { autoPlay } from 'react-swipeable-views-utils';

import * as S from './MobileCarousel.styles';

const AutoPlaySwipeableViews = autoPlay(SwipeableViews);

const defaultStyles = {
  slideContainer: {
    padding: '0 .5rem'
  }
};

export const MobileCarousel = ({
  carouselStyles = {},
  children,
  className,
  hasAutoPlay = false,
  hasProgressButtons = true,
  progressClassName
}: Props) => {
  const [current, setCurrent] = React.useState(0);

  const styles = { ...defaultStyles, ...carouselStyles };

  const SwipeableContainer = hasAutoPlay ? AutoPlaySwipeableViews : SwipeableViews;

  return (
    <S.MobileCarousel className={className}>
      <SwipeableContainer
        index={current}
        onChangeIndex={setCurrent}
        style={styles.root}
        slideStyle={styles.slideContainer}
      >
        {children}
      </SwipeableContainer>
      {hasProgressButtons && (
        <S.Progress className={progressClassName}>
          {children.map((child, i) => (
            <S.Button
              key={`carousel-${child.key}`}
              isCurrent={i === current}
              onClick={() => setCurrent(i)}
              title={`Slide ${i}`}
            >
              <S.Dot />
            </S.Button>
          ))}
        </S.Progress>
      )}
    </S.MobileCarousel>
  );
};

type Props = {
  carouselStyles?: {
    root?: any;
    slideContainer?: any;
  };
  children: React.ReactElement[];
  className?: string;
  hasAutoPlay?: boolean;
  hasProgressButtons?: boolean;
  progressClassName?: string;
};
