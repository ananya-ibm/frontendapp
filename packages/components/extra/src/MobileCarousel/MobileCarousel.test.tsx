/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import React from 'react';
import { act } from 'react-dom/test-utils';
import { cleanup, fireEvent, waitFor, screen } from '@testing-library/react';
import { renderWithTheme as render } from '@testUtils';
import { ContentCard } from '@exo/frontend-components-content';
import { axe } from 'jest-axe';
import { MobileCarousel } from './MobileCarousel';

afterEach(() => cleanup());

const testItems = [
  {
    title: 'New Offer 1',
    subtitle: 'Check out the latest offer',
    text: "Isn't it a great offer. I bet you've never seen one like it.",
    img:
      'https://images.unsplash.com/photo-1476286768413-e7051cdb2179?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=60',
    linkText: 'Learn More',
    link: '#',
    hasAlwaysShadow: true
  },
  {
    title: 'New Offer 2',
    subtitle: 'Check out the latest offer',
    text: "Isn't it a great offer. I bet you've never seen one like it.",
    img:
      'https://images.unsplash.com/photo-1476286768413-e7051cdb2179?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=60',
    linkText: 'Learn More',
    link: '#',
    hasAlwaysShadow: true
  },
  {
    title: 'New Offer 3',
    subtitle: 'Check out the latest offer',
    text: "Isn't it a great offer. I bet you've never seen one like it.",
    img:
      'https://images.unsplash.com/photo-1476286768413-e7051cdb2179?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=60',
    linkText: 'Learn More',
    link: '#',
    hasAlwaysShadow: true
  }
];

describe('<MobileCarousel /> component', () => {
  test('its snapshot matches', () => {
    const { container } = render(
      <MobileCarousel>
        {testItems.map(item => (
          <div key={item.title} style={{ padding: '.5rem 0' }}>
            <ContentCard {...item} />
          </div>
        ))}
      </MobileCarousel>
    );

    expect(container).toBeDefined();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('it passes an axe accesibility check', async () => {
    let container;
    let results;
    await act(async () => {
      // eslint-disable-next-line prefer-destructuring
      container = render(
        <MobileCarousel>
          {testItems.map(item => (
            <div key={item.title} style={{ padding: '.5rem 0' }}>
              <ContentCard {...item} />
            </div>
          ))}
        </MobileCarousel>
      ).container;

      results = await axe(container);
    });

    expect(results).toHaveNoViolations();
  });

  test('clicking on a button changes the slide', () => {
    const { getByTitle, getByText } = render(
      <MobileCarousel>
        {testItems.map(item => (
          <div key={item.title} style={{ padding: '.5rem 0' }}>
            <ContentCard {...item} />
          </div>
        ))}
      </MobileCarousel>
    );

    const button = getByTitle('Slide 2');
    fireEvent.click(button);

    // buttons indexes 0-based whereas slide titles are 1-based
    const thirdSlide = getByText('New Offer 3');
    expect(thirdSlide).toBeDefined();
  });

  test('slides can change automatically with auto play', async () => {
    const { getByText } = render(
      <MobileCarousel hasAutoPlay>
        {testItems.map(item => (
          <div key={item.title} style={{ padding: '.5rem 0' }}>
            <ContentCard {...item} />
          </div>
        ))}
      </MobileCarousel>
    );

    await waitFor(() => {
      expect(getByText('New Offer 2')).toBeInTheDocument();
    });
  });

  test('progress buttons can be hidden', async () => {
    render(
      <MobileCarousel hasProgressButtons={false}>
        {testItems.map(item => (
          <div key={item.title} style={{ padding: '.5rem 0' }}>
            <ContentCard {...item} />
          </div>
        ))}
      </MobileCarousel>
    );

    const button1 = screen.queryByLabelText('Slide 1');
    const button2 = screen.queryByLabelText('Slide 2');
    const button3 = screen.queryByLabelText('Slide 3');

    expect(button1).toBeNull();
    expect(button2).toBeNull();
    expect(button3).toBeNull();
  });
});
