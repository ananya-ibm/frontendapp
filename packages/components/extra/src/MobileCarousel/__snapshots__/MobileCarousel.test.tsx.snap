// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MobileCarousel /> component its snapshot matches 1`] = `
.c2 {
  background-image: url(https://images.unsplash.com/photo-1476286768413-e7051cdb2179?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=60);
  background-position: center;
  background-size: cover;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 0.5rem;
  padding-top: 56.25%;
  width: 100%;
}

.c4 {
  font-size: 0.9rem;
  margin-bottom: 1rem;
}

.c5 {
  font-size: 0.9rem;
}

.c3 {
  margin-bottom: 0.5rem;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding-bottom: 1rem;
  text-align: center;
  box-shadow: 0.125rem 0.125rem 1rem rgba(0,0,0,0.1), -0.125rem -0.125rem 1rem rgba(0,0,0,0.1);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.5rem;
}

.c9 {
  background: #161616;
  border-radius: 50%;
  cursor: pointer;
  height: 60%;
  width: 60%;
}

.c7 {
  background: transparent;
  border: 0 none;
  cursor: pointer;
  height: 1.25rem;
  opacity: 0.4;
  -webkit-transition: 200ms opacity ease-in;
  transition: 200ms opacity ease-in;
  width: 1.25rem;
  z-index: 3;
  opacity: 1;
}

.c7 .c8 {
  background-color: #0062ff;
}

.c7:hover,
.c7:focus {
  outline: none;
}

.c10 {
  background: transparent;
  border: 0 none;
  cursor: pointer;
  height: 1.25rem;
  opacity: 0.4;
  -webkit-transition: 200ms opacity ease-in;
  transition: 200ms opacity ease-in;
  width: 1.25rem;
  z-index: 3;
}

.c10:hover,
.c10:focus {
  outline: none;
}

@media (min-width:672px) {
  .c0 {
    display: none;
  }
}

<div
  class="c0"
>
  <div
    style="overflow-x: hidden;"
  >
    <div
      class="react-swipeable-view-container"
      style="flex-direction: row; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
    >
      <div
        aria-hidden="false"
        data-swipeable="true"
        style="width: 100%; flex-shrink: 0; overflow: auto; padding: 0px .5rem;"
      >
        <div
          style="padding: .5rem 0px;"
        >
          <div
            class="c1 content-card"
          >
            <div
              class="c2"
            />
            <h4
              class="ContentCardstyles__Title-sc-16fj6qq-3 card-title"
            >
              New Offer 1
            </h4>
            <h5
              class="c3 card-subtitle"
            >
              Check out the latest offer
            </h5>
            <div
              class="c4 card-content"
            >
              Isn't it a great offer. I bet you've never seen one like it.
            </div>
            <a
              class="c5"
              href="#"
            >
              Learn More
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="c6"
  >
    <button
      class="c7"
      title="Slide 0"
    >
      <div
        class="c8 c9"
      />
    </button>
    <button
      class="c10"
      title="Slide 1"
    >
      <div
        class="c8 c9"
      />
    </button>
    <button
      class="c10"
      title="Slide 2"
    >
      <div
        class="c8 c9"
      />
    </button>
  </div>
</div>
`;
